pkg/providers/amifamily/ami.go

```diff
@@ -302,3 +303,48 @@ func (p *Provider) getRequirementsFromImage(ec2Image *ec2.Image) scheduling.Requ
 	requirements.Add(scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, architecture))
 	return requirements
 }
+
+func FilterAmiByRequirements(amis []AMI, requirements scheduling.Requirements) []AMI {
+	var results []AMI
+	for _, ami := range amis {
+		keys := requirements.Keys().Intersection(ami.Requirements.Keys())
+		if len(keys) == 0 {
+			continue //no overlap, let's check next ami
+		}
+		compatible := true
+		for key := range keys {
+			requirement := requirements.Get(key)
+			amiRequirement := ami.Requirements.Get(key)
+			if requirement.Operator() == v1.NodeSelectorOpDoesNotExist {
+				compatible = false //machine requires OpDoesNotExist, but ami provides. incompatible
+				break
+			}
+			if requirement.Operator() == v1.NodeSelectorOpIn && len(lo.Intersect(requirement.Values(), amiRequirement.Values())) == 0 {
+				compatible = false //machine requires OpIn, but ami has no one. incompatible
+				break
+			}
+			if requirement.Operator() == v1.NodeSelectorOpNotIn && len(lo.Intersect(requirement.Values(), amiRequirement.Values())) > 0 {
+				compatible = false //machine requires OpNotIn, but ami has overlapped. incompatible
+				break
+			}
+		}
+		if compatible {
+			results = append(results, ami)
+		}
+	}
+
+	if len(results) != 0 {
+		return results
+	}
+	return amis
+}
+
+func (ami AMI) GetLabelsFromRequirements() map[string]string {
```

Comment:
For the drift cycle.
As said in https://github.com/aws/karpenter/pull/3769#discussion_r1169399806
the drift cycle recovers the machine requirements from node labels.
ec2 tag is the only solution I found that I can label nodes.
The kubelet args "--node-labels" doesn't work because of some unknown reasons. Perhaps https://github.com/aws/karpenter-core/blob/ab4eccca6536fa89c751953cad2c0611c5a9018c/pkg/controllers/provisioning/provisioner.go#L362 has some hidden issue.